### PR TITLE
NXmx equipment component

### PR DESCRIPTION
**Pull request by phyy-nx**
`equipment_component` support was added to the NeXus NXtransformations base class in nexusformat/definitions#1182, as ported in from ImageCIF:
http://www.bernstein-plus-sons.com/software/CBF/doc/cif_img_1.8.4.html#_axis.equipment_component

Description: https://github.com/nexusformat/definitions/pull/1182#issuecomment-1227609869, quoted here:

> `equipment_component` is an ImageCIF field @yayahjb and I added a [while back](http://cci.lbl.gov/publications/download/CCN_2014_p19.pdf) to support grouping of axes together to form a single change-of-basis operation for a component of a detector.  I had thought this was already in NeXus as I've been using it in DIALS for some time, both when [writing](https://github.com/cctbx/cctbx_project/blob/master/xfel/swissfel/jf16m_cxigeom2nexus.py#L344) and [reading](https://github.com/cctbx/dxtbx/blob/main/src/dxtbx/format/nexus.py#L721-L745) NeXus files.  I only recently discovered that it wasn't actually in NXtransformations.

So `equipment_component` was used by FormatNexus.py but wasn't supported FormatNXmx.py.  This has lead to detector models with extra hierarchy levels.  Here's a demo, given the snippet `show_hierarchy_depth.py`:

```
import dxtbx, sys
img = dxtbx.load(sys.argv[1])

pg = img.get_detector().hierarchy()
while True:
  print(pg.get_name(), pg.get_local_origin())
  if pg.is_panel():
    break
  else:
    pg = pg[0]
```
Run like this:
```
libtbx.python show_hierarchy_depth.py `dials.data get lysozyme_JF16M_4img -q`/lyso009a_0087.JF07T32V01_master_4img.h5
```

Results on main:
```
 (0.0, 0.0, 0.0)
/entry/instrument/ELE_D0/transformations/AXIS_RAIL (0.0, 0.0, -97.83000183105469)
/entry/instrument/ELE_D0/transformations/AXIS_D0 (0.4290327180241036, 0.12102061135946585, 0.294)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0 (-75.1333334212039, -84.65967506456187, 0.0)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0M0 (-38.310714303741115, -61.113781401191716, 0.0)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0M0A0 (-38.58423053508638, -19.131359799099098, 0.0)
/entry/instrument/ELE_D0/ARRAY_D0Q0M0A0 (0.0, 0.0, 0.0)
```
Note the unnamed root level and the 7 levels in total.

Results on this branch:
```
/entry/instrument/ELE_D0/transformations/AXIS_D0 (0.4290327180241036, 0.12102061135946585, -97.53600183105469)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0 (-75.1333334212039, -84.65967506456187, 0.0)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0M0 (-38.310714303741115, -61.113781401191716, 0.0)
/entry/instrument/ELE_D0/transformations/AXIS_D0Q0M0A0 (-38.58423053508638, -19.131359799099098, 0.0)
/entry/instrument/ELE_D0/ARRAY_D0Q0M0A0 (0.0, 0.0, 0.0)

```
Now it's 5 levels by dropping the top level and combining `AXIS_RAIL` and `AXIS_D0`.  Note that dials.show on the h5 file before and after this change shows no difference in the final per-panel vectors.

Also adjusts the JF16M test since the hierarchy has shrunk from 7 to 5 levels.

This will invalidate any geometry files made on JF16M data since #455 was merged.  That affects a few XFEL folks, and is easily resolved.  I don't know if this affects Eiger data with multiple modules.  It might change the number of hierarchy levels, since the blank root level is dropped, but it wouldn't affect the final per-panel vectors.

### CODE REVIEW

**Review by graeme-winter — COMMENTED** [review]
Comment not approval; possibly even a request for changes

Wonder if you could simplify the grouping code by making use of `itertools.groupby` - or at least make it shorter....

Also: narrative comment explaining what the grouping is doing would probably help future you and future me work in this region...

**Comment by phyy-nx** [comment]
@graeme-winter groupby is amazing.  ready for re-review.  thanks! :)

File: src/dxtbx/nexus/__init__.py [hunk 1]
```diff
@@ -273,19 +275,33 @@ def get_dxtbx_detector(
...
-                        continue
+                pg: dxtbx.model.Detector | dxtbx.model.Panel | None = None
+
+                # Group any transformations together that share the same equipement_component
+                # to reduce the number of hierarchy levels
+                def equipment_component_key(dependency):
+                    return dependency.equipment_component
+
+                for _, transformation_group in groupby(
+                    reversed_dependency_chain, key=equipment_component_key
```
Comment (graeme-winter) [inline src/dxtbx/nexus/__init__.py:286]: @ndevenish commented offline that 

```
key=lambda dep: dep.equipment_component_key
```

would remove the need for the explicit function (which would also be in keeping with the functional nature of itertools)

**Review by graeme-winter — APPROVED** [review]
Code much easier to read with the `groupby` implementation - passed on suggestion for the use of a `lambda` function in place of teh explicit key function which I think would again represent a small improvement otherwise LGTM

File: newsfragments/561.feature [hunk 2]
```diff
@@ -0,0 +1 @@
+Add support for better grouping of detector modules in NeXus data. Note, expt files will not be backwards compatible for users of the JF16M detector at SwissFEL or the AGIPD detector at EuXFEL.
```
Comment (rjgildea) [inline newsfragments/561.feature:1]: > Note, expt files will not be backwards compatible
Could you expand on this, or what the suggested solution is? Does this mean 

```suggestion
NXmx: Add support for [`@equipment_compenent`](https://manual.nexusformat.org/classes/base_classes/NXtransformations.html#nxtransformations-axisname-equipment-component-attribute) for forming logical groupings of transformations to reduce the number of levels in the detector hierarchy. Note, expt files will not be backwards compatible for users of the JF16M detector at SwissFEL or the AGIPD detector at EuXFEL.
```

File: src/dxtbx/nexus/__init__.py [hunk 3]
```diff
@@ -14,6 +14,8 @@
 from dxtbx import flumpy
 from dxtbx.nexus.nxmx import units
 
+from itertools import groupby
```
Comment (rjgildea) [inline src/dxtbx/nexus/__init__.py:17]:
```suggestion
import itertools
```
A matter of personal preference, but I find it helps to make it clearer what `groupby` function you are calling later. Given that you only use `groupby` once in this file, I'm not sure there's much to be gained by `from module import thing` over `import module`.

**Review by rjgildea — APPROVED** [review]
Thanks for adding this, the changes look good, and I agree that the use of `itertools.groupby` helps improve things. I have added one comment that I think you might want to sort on `@equipment_component` before calling `itertools.groupby` as the latter [assumes that the input is sorted](https://docs.python.org/3/library/itertools.html#itertools.groupby).

If you can manage to mock up a simple test that doesn't depend on the external data file that would be appreciated, but don't let this block merging if you don't have the time (also happy to help out with this). Something along the lines of [this](https://github.com/dials/dxtbx/blob/a5c227c66777472f1521222cd1f958e1a0ca0992/tests/nexus/test_nxmx.py#L243-L278) might work - you would only need e.g. two transformations that are grouped into a single transformation to make a basic test of the code.

It might be beneficial to rename the `test_jf16M` test function something a little more verbose/descriptive, e.g. `test_jf16M_matches_expected_hierarchy` to describe what you're actually testing, rather than just generically "something to do with jf16M".

**Comment by phyy-nx** [comment]
@graeme-winter @rjgildea so I realized that I couldn't sort the reversed_dependency_chain since it is a chain and the order has meaning.  But also I realized that if equipment_component is present for some transformations but absent for others, then that would reorder things incorrectly when grouping the absent transformations together.  So I had to increase the complexity a bit by adding some extra smarts to the key function, which also meant I didn't want to use a lambda function.

The other requests were applied.  Thanks for the reviews!  I can wait another day to merge in case there is more feedback.

**Comment by rjgildea** [comment]
Yes, I realise now that in this case the input to groupby is by definition sorted via the chain of dependencies!

**Comment by phyy-nx** [comment]
Whoops forgot to add that test, I'll do that tomorrow

**Comment by graeme-winter** [comment]
> @graeme-winter @rjgildea so I realized that I couldn't sort the reversed_dependency_chain since it is a chain and the order has meaning. But also I realized that if equipment_component is present for some transformations but absent for others, then that would reorder things incorrectly when grouping the absent transformations together. So I had to increase the complexity a bit by adding some extra smarts to the key function, which also meant I didn't want to use a lambda function.

Grouping absent transformations - yup good point! However should probably also assert that individual `equipment_component`s are (i) unique and (ii) connected e.g. you don't have stages which are not part of `equipment_component=x` mid way through `x`

**Pull request merged**